pkg/server/ptypes/pipeline_ui.go

```diff
@@ -0,0 +1,408 @@
+	var rootId string
+	var rootNode *pb.UI_PipelineRunTreeNode
+
+	// Populate job index
+	for _, job := range jobs {
+		jobIdx[job.Id] = job
+	}
+
+	// Populate status report index
+	for _, statusReport := range statusReports {
+		if d := statusReport.GetDeploymentId(); d != "" {
+			statusReportIdx[d] = statusReport
+		} else if r := statusReport.GetReleaseId(); r != "" {
+			statusReportIdx[r] = statusReport
+		}
+	}
+
+	// Populate node index
+	for _, job := range jobs {
+		var step *pb.Pipeline_Step
+		var statusReport *pb.StatusReport
+
+		if stepOp := job.GetPipelineStep(); stepOp != nil {
+			step = stepOp.Step
+		} else {
+			// This job represents a nested or referenced pipeline
+			// invocation. It doesn’t encode all the necessary
+			// information to infer the step details, so we’ll
+			// create a placeholder step and fill in the details as
+			// we learn more from other jobs in the set.
+			step = &pb.Pipeline_Step{
+				Name:      job.Pipeline.Step,
+				DependsOn: []string{},
+				Kind:      &pb.Pipeline_Step_Pipeline_{},
+			}
+		}
+
+		// Look up the latest status report for this node
+		if d := job.Result.GetDeploy().GetDeployment(); d != nil {
+			statusReport = statusReportIdx[d.Id]
+		}
+		if r := job.Result.GetRelease().GetRelease(); r != nil {
+			statusReport = statusReportIdx[r.Id]
+		}
+
+		// Build the node (children will be populated later)
+		node := &pb.UI_PipelineRunTreeNode{
+			Step:               step,
+			Job:                &pb.Ref_Job{Id: job.Id},
+			State:              uiPipelineRunTreeNodeStateFromJob(job),
+			StartTime:          job.AckTime,
+			CompleteTime:       job.CompleteTime,
+			Application:        job.Application,
+			Workspace:          job.Workspace,
+			Result:             job.Result,
+			LatestStatusReport: statusReport,
+			Children: &pb.UI_PipelineRunTreeNode_Children{
+				Mode:  pb.UI_PipelineRunTreeNode_Children_SERIAL,
+				Nodes: []*pb.UI_PipelineRunTreeNode{},
+			},
+		}
+
+		// Register the node the index
+		nodeIdx[job.Id] = node
+	}
+
+	// Build graph from job `DependsOn` fields
+	for i, job := range jobIdx {
+		g.Add(i)
+
+		for _, j := range job.DependsOn {
+			if _, exists := jobIdx[j]; !exists {
+				// If the referenced job isn’t in the set then
+				// we can safely ignore it. It’s likely a
+				// task-releated ancillary job.
+				continue
+			}
+
+			g.Add(j)
+			g.AddEdge(j, i)
+		}
+	}
+
+	// Remove short cuts from the graph
+	g.TransitiveReduction()
+
+	// Find the root node
+	{
+		sorted := g.KahnSort()
+
+		id, ok := sorted[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not find root node")
+		}
+
+		node, ok := nodeIdx[id]
+		if !ok {
+			return nil, fmt.Errorf("could not find root node")
+		}
+
+		rootId = id
+		rootNode = node
+	}
+
+	var process func(string, *pb.UI_PipelineRunTreeNode) error
+
+	injectVirtualNode := func(inputId string, nextIds []string, outputNode *pb.UI_PipelineRunTreeNode) (string, *pb.UI_PipelineRunTreeNode, error) {
```

Comment:
It looks like `injectVirtualNode` and `processSubPipeline` might be nice to have as their own functions rather than defined inside this one, I think it makes sense to abstract them out on their own, what do you think @jgwhite ? This would make it nice to write tests for as well.

Reply:
I’ve refactored this all significantly so there are more discrete functions. I didn’t write tests for `injectVirtualNode` and `processSubPipeline` in the end because I suspect they will change, so I’m treating them as an implementation detail. Keen to hear your guidance on this though.

Reply:
I think that's fair, especially if the function they are used in is tested? There won't be 100% coverage but that seems like a good first pass.

Reply:
> especially if the function they are used in is tested?

Yeah, I think the public function has a good set of test cases.